### ISSUE TITLE
Linear Scaling on GPU

### ISSUE THREAD

**Issue by CrBillman** [tensorflow/serving#623]
Hi everyone, thanks in advance for taking a look.  I'm trying to set up tensorflow serving to deploy a fine-tuned inception_v3 model, and trying to improve the performance.  Specifically, I'm surprised at the scaling on the GPU.  I expect that the model should have sublinear scaling on the GPU with respect to the batch size, as they should be parallelized.  However, the runtime has a linear dependence on the batch size.  In the plot below, you can see the inception timing as a function of the batch size on the left.  On the right was a test I did to see if it was due to the overhead of sending the images data, so I created a model that receives and pre-processes images in the same way, but fills the output with random data.  As you can see, the time for this part of the process is much smaller than the time for the entire inference calculation (but it does scale linearly as I would expect).

![serving_scaling](https://user-images.githubusercontent.com/7571302/31684543-bc79bb4a-b345-11e7-876d-092fdae590b0.png)

I'm using tensorflow 1.3 and tensorflow-serving 1.3, with CUDA 8.0 and cuDNN 5.1.10 (I have restrictions on the AWS instance that I intend to use, so for the moment these choices are restricted).  I've made sure to compile with optimized build options `bazel build -c opt --copt=-msse4.1 --copt=-msse4.2 --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-O3 --config=cuda --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0"`  I'm using the tensorflow_model_server for deployment, and the batching parameters for this test were:

```
max_batch_size { value: 1000 }

batch_timeout_micros { value: 1000 }

num_batch_threads { value:  8}
```

I'm not sure if this is useful, but the pre-processing steps I'm performing are:

```
img_placeholder = tf.placeholder(tf.float32, shape=[None,
                                                    input_image_size,
                                                    input_image_size,
                                                    3], name="input_image")

processed_batch = tf.map_fn(lambda x: preprocess_for_eval(x,
                                    output_image_size,
                                    output_image_size,
                                    central_fraction=0.85),
                            img_placeholder)

centered_batch = slim.batch_norm(processed_batch, center=True, scale=True,
                                 is_training=False)
with slim.arg_scope(inception_v3_arg_scope()):
        _, end_points = inception_v3(centered_batch,
                                     num_classes=num_classes,
                                     is_training=False)
```

I'm running on Ubuntu 16.04, and am using a GeForce GTX 960.  I'm confident that the GPU is being used, as I get output like 
```
2017-10-17 14:24:48.601485: I external/org_tensorflow/tensorflow/core/common_runtime/gpu/gpu_device.cc:1045] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 960, pci bus id: 0000:01:00.0)
``` 
when I run the tensorflow_model_server.  It's hard to find benchmarks, so I'm not sure if the timing that I'm getting is reasonable (for a batch size of 20, it takes ~220 ms).  But it definitely seems like the scaling with respect to the batch size is off.  Is this behavior expected?  If not, is there anything else I can do to troubleshoot and optimize?

Thanks!

**Comment by kirilg** [comment]
I would recommend following the [batching guide](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md) to search for reasonable values to use. In particular, I think setting batch_timeout_micros to 1000 (1ms) may mean that you're not actually waiting long enough to batch requests together and you're still processing much smaller batches.

Looking at the "GPU: One Approach" section, you can try setting `batch_timeout_micros` to something high, say 1000000 while tuning `max_batch_size`, then tune `batch_timeout_micros` for tail latency.

In your graphs, is the x-axis `max_batch_size` or are you always setting max_batch_size=1000 and the x-axis represents something else?

**Comment by CrBillman** [comment]
Hey kirilg,
I should have clarified this in the first post.  The max_batch_size is always 1000, and the batch size is the number of images that I send through in a single request.  During this initial testing, I'm using the API in a strange way.  I'm sending a single request with my entire batch organized on the client side.
```
request.inputs['images'].CopyFrom(tf.contrib.util.make_tensor_proto(data, shape=data.shape))
```

  So I don't think that the batch_timeout_micros should be an issue, because the request comes in as a single tensor.

**Comment by gautamvasudevan** [comment]
Closing due to staleness. If this is still an issue, please file a new updated issue with current steps to reproduce the bug. If this is a question, please ask it on: 

https://stackoverflow.com/questions/tagged/tensorflow-serving

Thanks!
